src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go

```diff
@@ -99,31 +99,62 @@ func (a *StorageMinerActorCode_I) _slashDealsFromFaultReport(rt Runtime, sectorN
 
 	Release(rt, h, st)
 
-	// TODO: Send(StorageMarketActor, ProcessDealSlash)
+	processDealSlashParam := make([]actor.MethodParam, 2)
+	rt.SendCatchingErrors(&msg.InvocInput_I{
+		To_:     addr.StorageMarketActorAddr,
+		Method_: market.MethodProcessDealSlash,
+		Params_: processDealSlashParam,
+	})
+
 }
 
-// construct FaultReport
+func (a *StorageMinerActorCode_I) _slashPledgeForStorageFault(rt Runtime, sectorNumbers []sector.SectorNumber, faultType sector.StorageFaultType) {
+	h, st := a.State(rt)
+
+	affectedPower := block.StoragePower(0)
+	for _, sectorNo := range sectorNumbers {
+
+		utilizationInfo := st._getUtilizationInfo(rt, sectorNo)
+		affectedPower += utilizationInfo.CurrUtilization()
+
+	}
+
+	Release(rt, h, st)
+
+	panic(affectedPower)
+	slashPledgeParams := make([]actor.MethodParam, 3)
+
+	rt.SendCatchingErrors(&msg.InvocInput_I{
+		To_:     addr.StoragePowerActorAddr,
+		Method_: power.MethodSlashPledgeForStorageFault,
+		Params_: slashPledgeParams,
+	})
+}
+
+// should always happen after _submitPowerReport to have the most up to date utilization info
 // reset NewTerminatedFaults
-func (a *StorageMinerActorCode_I) _submitFaultReport(
+func (a *StorageMinerActorCode_I) _slashCollateralForStorageFaults(
 	rt Runtime,
-	newDeclaredFaults sector.CompactSectorSet,
-	newDetectedFaults sector.CompactSectorSet,
-	newTerminatedFaults sector.CompactSectorSet,
+	newDeclaredFaults sector.CompactSectorSet, // diff value
+	newDetectedFaults sector.CompactSectorSet, // diff value
+	newTerminatedFaults sector.CompactSectorSet, // diff value
 ) {
-	faultReport := &sector.FaultReport_I{
-		NewDeclaredFaults_:   newDeclaredFaults,
-		NewDetectedFaults_:   newDetectedFaults,
-		NewTerminatedFaults_: newTerminatedFaults,
+
+	// only terminatedFault will result in collateral deal slashing
+	if len(newTerminatedFaults) > 0 {
+		a._slashDealsForStorageFault(rt, newTerminatedFaults.SectorsOn(), sector.TerminatedFault)
+		a._slashPledgeForStorageFault(rt, newTerminatedFaults.SectorsOn(), sector.TerminatedFault)
```

Comment:
Do these methods need to take `faultType` as an argument? I assume this could just be `_slashDeals(...)`, `_slashPledge(...)` as the behavior is independent of the original cause.

Reply:
that is fine, we will move the logic of computing the amount to slash else where. In `_slashPledge` for example, we expect different amount/proportion depending on the fault type.

Reply:
Ah, I don't have a strong preference on this; was just wondering, since those flags seemed to be unused in the current PR code. Wherever you think best to implement those logic operations sounds good to me.